### PR TITLE
[WIP] Feature/rebasing

### PR DESCRIPTION
**Pull request by glvr182**
@ all feel free to add ideas, tips, comments to this pr

### CODE REVIEW

**Comment by glvr182** [comment]
Rebased it onto the latest master using the rebasing feature :D

File: pkg/gui/branches_panel.go [hunk 1]
```diff
@@ -10,6 +10,27 @@ import (
...
+	title := "Rebasing"
+	prompt := fmt.Sprintf("Are you sure you want to rebase %s onto %s?", checkedOutBranch, selectedBranch)
+
+	return gui.createConfirmationPanel(g, v, title, prompt,
+		func(g *gocui.Gui, v *gocui.View) error {
+			if selectedBranch == checkedOutBranch {
+				return gui.createErrorPanel(g, gui.Tr.SLocalize("CantRebaseOntoSelf"))
+			}
+			if err := gui.GitCommand.RebaseBranch(selectedBranch); err != nil {
+				gui.createErrorPanel(g, "Failed to rebase")
```
Comment (golangcibot) [inline pkg/gui/branches_panel.go]: Error return value of `gui.createErrorPanel` is not checked (from `errcheck`)

File: pkg/gui/branches_panel.go [hunk 2]
```diff
@@ -10,6 +10,30 @@ import (
...
+			if selectedBranch == checkedOutBranch {
+				return gui.createErrorPanel(g, gui.Tr.SLocalize("CantRebaseOntoSelf"))
+			}
+			if err := gui.GitCommand.RebaseBranch(selectedBranch); err != nil {
+				if err := gui.createErrorPanel(g, err.Error()); err != nil {
+					gui.Log.Error(err.Error())
+				}
+			}
+
+			gui.GitCommand.AbortRebaseBranch()
```
Comment (golangcibot) [inline pkg/gui/branches_panel.go]: Error return value of `gui.GitCommand.AbortRebaseBranch` is not checked (from `errcheck`)

File: pkg/gui/branches_panel.go [hunk 3]
```diff
@@ -10,6 +10,35 @@ import (
...
+	prompt := fmt.Sprintf("Are you sure you want to rebase %s onto %s?", checkedOutBranch, selectedBranch)
+
+	return gui.createConfirmationPanel(g, v, title, prompt,
+		func(g *gocui.Gui, v *gocui.View) error {
+			if selectedBranch == checkedOutBranch {
+				return gui.createErrorPanel(g, gui.Tr.SLocalize("CantRebaseOntoSelf"))
+			}
+
+			if err := gui.GitCommand.RebaseBranch(selectedBranch); err != nil {
+				if err := gui.createConfirmationPanel(g, v, "Rebase failed", "Rebasing failed, would you like to resolve it?",
```
Comment (golangcibot) [inline pkg/gui/branches_panel.go]: empty branch (from `megacheck`)

File: pkg/gui/merge_panel.go [hunk 4]
```diff
@@ -259,5 +259,9 @@ func (gui *Gui) handleCompleteMerge(g *gocui.Gui) error {
 	}
 	gui.stageSelectedFile(g)
 	gui.refreshFiles(g)
+	if rebase, err := gui.GitCommand.IsInRebaseState(); rebase && err == nil {
+		gui.GitCommand.ContinueRebaseBranch()
```
Comment (golangcibot) [inline pkg/gui/merge_panel.go]: Error return value of `gui.GitCommand.ContinueRebaseBranch` is not checked (from `errcheck`)

File: pkg/gui/merge_panel.go [hunk 5]
```diff
@@ -259,5 +259,9 @@ func (gui *Gui) handleCompleteMerge(g *gocui.Gui) error {
...
+		gui.refreshSidePanels(g)
```
Comment (golangcibot) [inline pkg/gui/merge_panel.go]: Error return value of `gui.refreshSidePanels` is not checked (from `errcheck`)

**Comment by glvr182** [comment]
This about works, just need to test it more to find edge cases. Also I should create a unit test

**Comment by glvr182** [comment]
@jesseduffield Can you help me test?

File: pkg/gui/merge_panel.go [hunk 6]
```diff
@@ -241,20 +253,45 @@ func (gui *Gui) renderMergeOptions() error {
...
+			if err := gui.genericRebaseCommand("continue"); err != nil {
+				if err == gui.Errors.ErrSubProcess {
+					return err
+				}
+				if strings.Contains(err.Error(), "No changes - did you forget to use") {
+					if err := gui.genericRebaseCommand("skip"); err != nil {
+						if err == gui.Errors.ErrSubProcess {
+							return err
+						}
+						gui.createErrorPanel(gui.g, err.Error())
```
Comment (golangcibot) [inline pkg/gui/merge_panel.go]: Error return value of `gui.createErrorPanel` is not checked (from `errcheck`)

File: pkg/gui/merge_panel.go [hunk 7]
```diff
@@ -241,20 +253,45 @@ func (gui *Gui) renderMergeOptions() error {
...
+					}
+				} else {
+					gui.createErrorPanel(gui.g, err.Error())
```
Comment (golangcibot) [inline pkg/gui/merge_panel.go]: Error return value of `gui.createErrorPanel` is not checked (from `errcheck`)

File: pkg/gui/merge_panel.go [hunk 8]
```diff
@@ -241,20 +261,44 @@ func (gui *Gui) renderMergeOptions() error {
...
+	return nil
 }
 
-func (gui *Gui) handleCompleteMerge(g *gocui.Gui) error {
-	filesView, err := g.View("files")
-	if err != nil {
-		return err
+func (gui *Gui) handleCompleteMerge() error {
+	filesView := gui.getFilesView()
+	gui.stageSelectedFile(gui.g)
```
Comment (golangcibot) [inline pkg/gui/merge_panel.go]: Error return value of `gui.stageSelectedFile` is not checked (from `errcheck`)

File: pkg/commands/git.go [hunk 9]
```diff
@@ -502,24 +543,105 @@ func (c *GitCommand) getMergeBase() (string, error) {
...
+	rewrittenCount := 0
+	bytesContent, err := ioutil.ReadFile(".git/rebase-apply/rewritten")
+	if err == nil {
+		content := string(bytesContent)
+		rewrittenCount = len(strings.Split(content, "\n"))
+	}
+
+	// we know we're rebasing, so lets get all the files whose names have numbers
+	commits := []*Commit{}
+	err = filepath.Walk(".git/rebase-apply", func(path string, f os.FileInfo, err error) error {
```
[the remark this answers is not in the document]
Comment (glvr182) [reply pkg/commands/git.go]: And you could change it to 
```
if err := filepath.Walk(); err != nil{}
```
But its your call

File: pkg/commands/os.go [hunk 10]
```diff
@@ -202,3 +202,14 @@ func (c *OSCommand) CreateTempFile(filename, content string) (string, error) {
 func (c *OSCommand) RemoveFile(filename string) error {
 	return os.Remove(filename)
 }
+
+// FileExists checks whether a file exists at the specified path
+func (c *OSCommand) FileExists(path string) (bool, error) {
```
Comment (glvr182) [inline pkg/commands/os.go]: Isn't there a function for this already in the standard library?

File: pkg/commands/os.go [hunk 11]
```diff
@@ -202,3 +202,14 @@ func (c *OSCommand) CreateTempFile(filename, content string) (string, error) {
...
+	if _, err := os.Stat(path); err != nil {
+		if os.IsNotExist(err) {
```
Comment (glvr182) [inline pkg/commands/os.go]: This one
